corehq/project_limits/rate_limiter.py

```diff
@@ -106,8 +113,8 @@ def wait(self, scope, timeout, windows_not_to_wait_on=('hour', 'day', 'week')):
                 time.sleep(delay)
 
 
-@quickcache(['domain'], memoize_timeout=60, timeout=60 * 60)
-def get_n_users_for_rate_limiting(domain):
+@quickcache(['domain', 'enterprise_limit'], memoize_timeout=60, timeout=60 * 60)
+def get_n_users_for_rate_limiting(domain, enterprise_limit=False):
```

Comment:
I'm trying to follow `get_n_users_for_rate_limiting` and the helper function it calls, `_get_users_included_in_subscription`, and finding it a little hard (1) to know what they're _supposed_ to be doing and (2) know that it's doing it right.

I think (and tell me if this is wrong) basically this function is being used to answer two different questions: "How many users belong to this domain" (if `enterprise_limit=False`) and "How many users belong to this enterprise account" (if `enterprise_limit=True`).

The way I think those should be calculated (and which it's hard for me to tell if this is doing) is

- "How many users belong to this domain": the number of actual users in the domain
- "How many users belong to this software plan / enterprise account": the number of users in the software plan the domain is linked to (regardless of whether it's a multi-domain enterprise account or just the software plan for the single domain)

Note that the current method, for single domains not a part of an enterprise account, is to take the max of the actual number of users and the number of users in the software plan. What I just described is a different way of doing that that amounts to the same thing for this special case, but also generalizes well to the multi-domain enterprise account case. This is because for a counter to be considered over its limit, it will have to be over _both_ the limit associated with the real number of users _and_ the limit associated with the number of users in the software plan, which is equivalent to it being over the limit associated with the greater (max) of the two.

You might have implemented it a little differently here, and it may even be correct / amount to the same thing ,but it's hard for me to follow with the different branching. I'd recommend just having two different functions `get_n_users_in_domain(domain)` and `get_n_users_in_subscription(domain)` and have them return their respective simple values (no max or branching needed). I'm also including a comment below showing how I'd then call it.

Reply:
This was also the part that was most confusing to me.

For the enterprise pool, do all projects in the enterprise share the same pool?

Reply:
Check out the bottom section of
> Original design doc https://docs.google.com/document/d/1a-ELOcCgHSFOFXjAN-ZhbVgxz4kK2yp9E_6PK5Fn2ZA/edit#

for a fuller description of the plan Cal and I both have in mind, but the basic description is this. There are three levels of counters: global, account, and domain. Each time an event comes in it increments all three (the global one, the account for the domain, and the domain). A domain's events are considered over the limit only when _all three_ relevant counters are over their limits. This differs from our current setup by its addition of the middle tier (the account level) which had been sort of clumsily fused into the domain level for single-domain accounts by taking the max of users in subscription and actual users in domain, a fusion that really did not work well at all for accounts with multiple domains. But yeah the section in that doc goes into just a little more detail.

The way Cal's implementing that here is to combine account and domain into one counter, and keep the global one separate, which I think seems reasonable.